### PR TITLE
Add support for integer path segments in assocPath

### PR DESCRIPTION
**Pull request by scott-christopher**
This is an alternative proposal to #1924 for supporting integer path segments in `assocPath` (and `lensPath` by its current implementation).

### CODE REVIEW

File: src/assocPath.js [hunk 1]
```diff
@@ -27,12 +31,19 @@ var assoc = require('./assoc');
...
+  if (path.length === 0) {
+    return val;
+  }
+  var idx = path[0];
+  if (path.length > 1) {
+    var nextObj = _has(idx, obj) ? obj[idx] : _isInteger(path[1]) ? [] : {};
+    val = assocPath(_slice(path, 1), val, nextObj);
+  }
+  if (_isInteger(idx) && _isArray(obj)) {
+    var arr = [].concat(obj);
```
Comment (scott-christopher) [inline src/assocPath.js]: Would be nice to use `update` here instead, though if we want to continue supporting the creation of intermediate paths when they don't exist in the target object then we need to allow for the creation of sparse arrays (which `update` prevents).

Comment (buzzdecafe) [inline src/assocPath.js]: yes, sparse array creation is what i didn't like about this in prior discussions. Current lensIndex behavior does not update non-existent indices

**Comment by CrossEye** [comment]
Since @buzzdecafe [set me straight][mh], I would definitely prefer this approach.

Any objections?


  [mh]: https://github.com/ramda/ramda/pull/1924#issuecomment-250798430

**Comment by buzzdecafe** [comment]
not enough to block it. i fret about the sparse array thing. that is pretty gross. this impl is less surprising, and maybe that is enough to justify it.

**Comment by jgoux** [comment]
Yes, yes and yes!
Can't wait to use it! :star2: 
Thanks @scott-christopher

**Comment by buzzdecafe** [comment]
oops, this one slipped between the cracks. i think it's good to merge, but needs some conflict resolution

**Comment by CrossEye** [comment]
Agreed.  @scott-christopher: do you have time to address the merge issue?

**Comment by scott-christopher** [comment]
Sure, I should be able to take a look later this afternoon.

**Comment by scott-christopher** [comment]
The conflict was just due to the change from `_slice(path, 1)` to `Array.prototype.slice.call(path, 1)` :zap:

**Comment by buzzdecafe** [comment]
:cow2:

**Comment by CrossEye** [comment]
:herb:

Very glad to have this change in!

**Comment by jgoux** [comment]
Congrats! :sheep: 
Would it be possible to tag a release/publish it to npm?

**Comment by buzzdecafe** [comment]
> Very glad to have this change in!

more equivocal. but maybe we can button up the loose ends later

**Comment by diedsmiling** [comment]
👍 Thanks for implementing this feature. I need very much at my project. When would be a release to npm?

**Comment by buzzdecafe** [comment]
still a few things to iron out before next release; gotta herd the cats

**Pull request merged**